Add handling for indexed-color image type in WP Image Editor

pbearne opened the pull request:
A new functionality has been added to the WP Image Editor class to check if an image is indexed-color encoded. This property is considered when saving a resized image to preserve its original image type. This reduces instances where the resized image filesize would end up larger than the original. A test to verify smaller file size for resized images has also been included.


Trac ticket: https://core.trac.wordpress.org/ticket/36477

Commit:
Add handling for indexed-color image type in WP Image Editor

A new functionality has been added to the WP Image Editor class to check if an image is indexed-color encoded. This property is considered when saving a resized image to preserve its original image type. This reduces instances where the resized image filesize would end up larger than the original. A test to verify smaller file size for resized images has also been included.

Commit:
Add handling for indexed-color image type in WP Image Editor

A new functionality has been added to the WP Image Editor class to check if an image is indexed-color encoded. This property is considered when saving a resized image to preserve its original image type. This reduces instances where the resized image filesize would end up larger than the original. A test to verify smaller file size for resized images has also been included.

mukeshpanchal27 commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -21,6 +21,15 @@ class WP_Image_Editor_Imagick extends WP_Image_Editor {
 	 */
 	protected $image;
 
+	/**
+     * Stores the information whether the image is indexed-color encoded
+     *
+     * @since 4.7
+     * @access protected
+     * @var bool
+     */
```
```suggestion
	/**
	 * Stores the information whether the image is indexed-color encoded
	 *
	 * @since 4.7
	 * @access protected
	 * @var bool
	 */
```

pbearne committed:
Update src/wp-includes/class-wp-image-editor-imagick.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Commit:
Refactor image resize test to use data provider

The image resizing test in `editorImagick.php` was refactored to utilize a data provider. This allows for a more comprehensive testing with a variety of 16-bit image files. Now, multiple test files including `17-c3-duplicate-entries.png`, `cloudflare-status.png`, and more are considered in the test case.

nosilver4u committed:
detect indexed color PNG images and quantize after resizing

nosilver4u committed:
combine alpha code and make png chunks conditional

nosilver4u committed:
fix palette image type condition

nosilver4u committed:
prevent grayscale conversion and only preserve tRNS for alpha

nosilver4u committed:
reduce the max colors (no buffer)

The buffer only seemed to help on the image that was getting converted to grayscale. But once the grayscale issue was fixed, the buffer wasn't having the desired effect on quality.

Commit:
Merge branch 'refs/heads/trunk' into 36477-keep-color-depth

Commit:
Merge remote-tracking branch 'refs/remotes/nosilver4u/png8-resize' into 36477-keep-color-depth

# Conflicts:
#	src/wp-includes/class-wp-image-editor-imagick.php

Commit:
end of day

Commit:
Merge branch 'refs/heads/trunk' into 36477-keep-color-depth

Commit:
Refactor png color depth test and clean debug code

Optimized the png color depth with valid png unit test by adding the expected_encoded parameter to match the expected result. Additionally, debugging code has been removed from WP_Image_Editor_Imagick, improving its readability and performance.
